### PR TITLE
Generate a proper service certificate for the autoscaler service.

### PR DESCRIPTION
**Pull request by markusthoemmes**
Today, the autoscaler's custom-metrics API will start to fail running a few hours on an Openshift cluster. That's likely due to some form of certificate rotation breaking it.

Based on the documentation for the prometheus adapter (which is another implementation of the custom-metrics API) here: https://docs.openshift.com/container-platform/4.1/monitoring/exposing-custom-application-metrics-for-autoscaling.html this generates proper serving certificates for the API server to use and hooks them up in the existing manifests.

### CODE REVIEW

**Comment by nak3** [comment]
LGTM :+1:  I have quickly tested it and looks all fine.

```
$ oc get svc -n knative-serving autoscaler -o yaml |grep annotation -A4
  annotations:
    manifestival: new
    service.alpha.openshift.io/serving-cert-secret-name: autoscaler-adapter-tls
    service.alpha.openshift.io/serving-cert-signed-by: openshift-service-serving-signer@1562305585
    service.beta.openshift.io/serving-cert-signed-by: openshift-service-serving-signer@1562305585

$ oc get secret -n knative-serving autoscaler-adapter-tls
NAME                     TYPE                DATA   AGE
autoscaler-adapter-tls   kubernetes.io/tls   2      41m

$ oc set volumes  -n knative-serving  pod/autoscaler-57cbf9fbb9-95vrb
pods/autoscaler-57cbf9fbb9-95vrb
  configMap/config-autoscaler as config-autoscaler
    mounted at /etc/config-autoscaler
  configMap/config-logging as config-logging
    mounted at /etc/config-logging
  configMap/config-observability as config-observability
    mounted at /etc/config-observability
  secret/autoscaler-adapter-tls as volume-serving-cert
    mounted at /var/run/serving-cert
  secret/controller-token-4z9zn as controller-token-4z9zn
    mounted at /var/run/secrets/kubernetes.io/serviceaccount

$ oc get pod -n knative-serving  autoscaler-57cbf9fbb9-95vrb -o yaml |grep args -A3
  - args:
    - --secure-port=8443
    - --tls-cert-file=/var/run/serving-cert/tls.crt
    - --tls-private-key-file=/var/run/serving-cert/tls.key
```

File: pkg/controller/knativeserving/openshift/openshift.go [hunk 1]
```diff
@@ -372,3 +376,66 @@ func ensureNamespace(c client.Client, ns string) error {
 	}
 	return nil
 }
+
+// annotateAutoscalerService annotates the autoscaler service with an Openshift annotation
+// that causes it to generate a certificate for the cluster to use internally.
+// Adapted from: https://docs.openshift.com/container-platform/4.1/monitoring/exposing-custom-application-metrics-for-autoscaling.html
+func annotateAutoscalerService(u *unstructured.Unstructured) error {
+	const annotationKey = "service.alpha.openshift.io/serving-cert-secret-name"
```
[the remark this answers is not in the document]
Comment (markusthoemmes) [reply pkg/controller/knativeserving/openshift/openshift.go]: Hmm, I opted to only define constants that are used across multiple steps there. I'd leave the specific ones where they are.

File: pkg/controller/knativeserving/openshift/openshift.go [hunk 2]
```diff
@@ -372,3 +376,66 @@ func ensureNamespace(c client.Client, ns string) error {
...
+		u.SetAnnotations(annotations)
+	}
+	return nil
+}
+
+// augmentAutoscalerDeployment mounts the secret generated by 'annotateAutoscalerService' into
+// the autoscaler deployment and makes sure the custom-metrics API uses the mounted certs properly.
+func augmentAutoscalerDeployment(u *unstructured.Unstructured) error {
+	const volumeName = "volume-serving-cert"
+	const mountPath = "/var/run/serving-cert"
```
[the remark this answers is not in the document]
Comment (markusthoemmes) [reply pkg/controller/knativeserving/openshift/openshift.go]: same as above. These are very specific constants we don't need to have globally available I think.

**Comment by markusthoemmes** [comment]
Thank you @nak3 and @savitaashture. I'll wait for QE to verify this works as intended and then we can merge it.

File: pkg/controller/knativeserving/openshift/openshift.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (savitaashture) [reply pkg/controller/knativeserving/openshift/openshift.go]: Okey.
Then here we can group const
```
const (
    volumeName = "volume-serving-cert"
    mountPath = "/var/run/serving-cert"
)

**Pull request merged**